### DISCUSSION TITLE
Global 'strictQuery' is ignored when schema/models are defined before mongoose setup

### DISCUSSION THREAD

**Discussion by Maximusya** [Automattic/mongoose#11362] [Q&A] [answered]
There are many issues like https://github.com/Automattic/mongoose/issues/11048 regarding `strictQuery` option. Every issue seem to be closed with the resolution to use `mongoose.set('strictQuery', false)`.

What we noticed in out project after migrating to mongoose 6 is that `mongoose.set('strictQuery', false)` seems to be ignored.
I boiled the difference between samples and our project structure down to the following:
- when schema/module definition is in the same module as setup and AFTER mongoose setup and connection, it works as per docs.
- when schema/module definition is in the same module as setup but BEFORE mongoose setup and connection, `strictQuery` is ignored.
- when schema/module is imported from separate module, `strictQuery` is ignored.
- `.setOptions({ strictQuery: false })` query modifier correctly turns off sanitizing in any case.

> $ node single-module-postdeclare.mjs 
Mongoose: users.insertOne({ name: 'foo', _id: new ObjectId("620279042947fba5613d525d"), __v: 0 }, { session: null })
Mongoose: users.findOne({ wrongField: 'bar' }, { projection: {} })
user found:  false

> $ node single-module-predeclare.mjs 
Mongoose: users.insertOne({ name: 'foo', _id: new ObjectId("62036971365c03a39e1ccd91"), __v: 0 }, { session: null })
Mongoose: users.findOne({}, { projection: {} })
user found:  true

> $ node split-modules.mjs 
Mongoose: users.insertOne({ name: 'foo', _id: new ObjectId("6202782cdb849318cc33547e"), __v: 0 }, { session: null })
Mongoose: users.findOne({}, { projection: {} })
user found:  true

What is wrong with this schema definition and mongoose setup order?..

## Single (setup first, schema second)

*single-module-postdeclare.mjs*
```javascript
import mongoose from 'mongoose';

run().catch(console.error);

async function run() {
  try {
    mongoose.set('debug', true);
    mongoose.set('strictQuery', false);
    await mongoose.connect('mongodb://localhost:27017/test');

    const userSchema = new mongoose.Schema({ name: String });
    const User = mongoose.model('User', userSchema);

    await User.create({ name: 'foo' });
    const user = await User.findOne({ wrongField: 'bar' })
      // .setOptions({ strictQuery: false });

    console.log('user found: ', user != null);
  } finally {
    await mongoose.disconnect();
  }
}
```

> $ node single-module-postdeclare.mjs 
Mongoose: users.insertOne({ name: 'foo', _id: new ObjectId("620279042947fba5613d525d"), __v: 0 }, { session: null })
Mongoose: users.findOne({ wrongField: 'bar' }, { projection: {} })
user found:  false

## Single (schema first, setup second)

*single-module-predeclare.mjs*
```javascript
import mongoose from 'mongoose';

run().catch(console.error);

async function run() {
  try {
    const userSchema = new mongoose.Schema({ name: String });
    const User = mongoose.model('User', userSchema);

    mongoose.set('debug', true);
    mongoose.set('strictQuery', false);
    await mongoose.connect('mongodb://localhost:27017/test');

    await User.create({ name: 'foo' });
    const user = await User.findOne({ wrongField: 'bar' })
      // .setOptions({ strictQuery: false });

    console.log('user found: ', user != null);
  } finally {
    await mongoose.disconnect();
  }
}
```

> $ node single-module-predeclare.mjs 
Mongoose: users.insertOne({ name: 'foo', _id: new ObjectId("62036971365c03a39e1ccd91"), __v: 0 }, { session: null })
Mongoose: users.findOne({}, { projection: {} })
user found:  true

## Split

*split-modules.js*
```javascript
import mongoose from 'mongoose';
import User from './User.mjs';

run().catch(console.error);

async function run() {
  try {
    mongoose.set('debug', true);
    mongoose.set('strictQuery', false);
    await mongoose.connect('mongodb://localhost:27017/test');

    await User.create({ name: 'foo' });
    const user = await User.findOne({ wrongField: 'bar' })
      // .setOptions({ strictQuery: false });

    console.log('user found: ', user != null);
  } finally {
    await mongoose.disconnect();
  }
}
```

*User.mjs*
```javascript
import mongoose from 'mongoose';

const userSchema = new mongoose.Schema({ name: String });
const User = mongoose.model('User', userSchema);

export default User;
```

> $ node split-modules.mjs 
Mongoose: users.insertOne({ name: 'foo', _id: new ObjectId("6202782cdb849318cc33547e"), __v: 0 }, { session: null })
Mongoose: users.findOne({}, { projection: {} })
user found:  true

**Comment by vkarpov15** [comment]
I'd say this is by design. Put `mongoose.set('strictQuery', false)` in your program's entry point.

**Reply by Maximusya** [reply]
1. What exactly is by design?
2. Every example has `mongoose.set('strictQuery', false)` in the entry point

Can you please elaborate or point to the docs?

**Reply by vkarpov15** [reply]
Right now, if you change global options _after_ compiling a model using `mongoose.model()`, the global option changes don't propagate to the model you compiled.

**Reply by Maximusya** [reply]
Thanks!
Yes, once I imported "db"-module before any "model"-module in the entry module, global options are correctly picked up.
For the reference I've put a sample working code structure in the answer.

**Comment by Maximusya** [comment, marked as answer]
Thanks, @vkarpov15!

The following code structure seem to put things in the correct order and all the desired global mongoose options are honored by the model:
> $ node index.js 
Mongoose: users.findOne({ wrongField: 'bar' }, { projection: {} })

*index.js*:
```javascript
import { db } from './db.js';
import User from './User.js';

run().catch(console.error);

async function run() {
  try {
    await db.connect();
    const user = await User.findOne({ wrongField: 'bar' })
  } finally {
    await db.disconnect();
  }
}
```

*db.js*:
```javascript
import mongoose from 'mongoose';

mongoose.set('debug', true);
mongoose.set('strictQuery', false);
mongoose.set('autoCreate', false);
mongoose.set('autoIndex', false);

export const db = {
  connect: () => mongoose.connect('mongodb://localhost:27017/test'),
  disconnect: () => mongoose.disconnect()
};
```

*User.js*:
```javascript
import mongoose from 'mongoose';

const userSchema = new mongoose.Schema({ name: String });
const User = mongoose.model('User', userSchema);

export default User;
```
